pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -140,41 +142,61 @@ type reconciler struct {
 	volumePluginMgr               *volumepkg.VolumePluginMgr
 	kubeletPodsDir                string
 	timeOfLastSync                time.Time
+	volumesFailedReconstruction   []podVolume
+	volumesNeedDevicePath         []v1.UniqueVolumeName
 }
 
 func (rc *reconciler) Run(stopCh <-chan struct{}) {
-	wait.Until(rc.reconciliationLoopFunc(), rc.loopSleepDuration, stopCh)
-}
-
-func (rc *reconciler) reconciliationLoopFunc() func() {
-	return func() {
-		rc.reconcile()
-
-		// Sync the state with the reality once after all existing pods are added to the desired state from all sources.
-		// Otherwise, the reconstruct process may clean up pods' volumes that are still in use because
-		// desired state of world does not contain a complete list of pods.
-		if rc.populatorHasAddedPods() && !rc.StatesHasBeenSynced() {
-			klog.InfoS("Reconciler: start to sync state")
-			rc.sync()
-		}
-	}
+	rc.reconstructVolumes()
```

Comment:
Since reconstructVolumes() is executed before reconciler even starts, we need to make sure it does not block. Checking he code there, so far seems fine. There are a few sys calls like os.Open(), I only see issue related to open named pipefile cause process hang. https://github.com/golang/go/issues/33050

In case the large number of existing mounts, it might cause slight delay on reconciler.

Reply:
In theory, any I/O can block when the storage backend is not available (like /var/lib/kubelet is on a NFS server or a network disk). In practice, if `open()` blocks, then the machine has more serious problems than reconstruction.

I went through all volume plugins and none of them seem to issue any API calls. They're parsing mount tables, json files and similar stuff. A second pair of eyes is welcome!

Reply:
iirc some NFS volumes can cause permanent locks in case nfs-server was not available.